search/src/Filtering.cpp

```diff
@@ -46,5 +47,207 @@ std::vector<int> sigmaGFilteredIndices(const std::vector<float>& values,
+                                                                         std::vector<double> vars,
+                                                                         std::vector<int> fluxIdx, bool reverse)
+{
+    int fluxSize = fluxValues.size();
+    double kalmanFluxes[fluxSize];
+    
+    std::vector<double> xhat(fluxSize, 0.0);
+    std::vector<double> p(fluxSize, 0.0);
+    double xhatMinus[fluxSize];
+    double pMinus[fluxSize];
+    double k[fluxSize];
+    
+    double q = 1.0;
+    
+    /* If reverse is false, calculate
+       Kalman result going 'forward'
+       (0 -> end of fluxes)
+       Else, calculate it 'backwards'
+       (end of fluxes -> 0)*/
+    if(!reverse) {
+        xhat[fluxIdx[0]] = fluxValues[fluxIdx[0]];
+        p[fluxIdx[0]] = vars[fluxIdx[0]];
+        
+        for(int i = 1; i < fluxIdx.size(); i++) {
+            int idx = fluxIdx[i];
+            int idxMinus = fluxIdx[i - 1];
+            
+            xhatMinus[idx] = xhat[idxMinus];
+            pMinus[idx] = p[idxMinus] + q;
+
+            k[idx] = pMinus[idx] / (pMinus[idx] + vars[idx]);
+            xhat[idx] = xhatMinus[idx] + k[idx]*(fluxValues[idx]-xhatMinus[idx]);
+            p[idx] = (1.0-k[idx])*pMinus[idx];
+        }
+    }
+    else {
+        xhat[fluxIdx.back()] = fluxValues[fluxIdx.back()];
+        p[fluxIdx.back()] = vars[fluxIdx.back()];
+        
+        for(int j = fluxIdx.size() - 2; j >= 0; j--) {
+            int idx = fluxIdx[j];
+            xhatMinus[idx] = xhat[fluxIdx[j + 1]];
+            pMinus[idx] = p[fluxIdx[j + 1]] + q;
+
+            k[idx] = pMinus[idx] / (pMinus[idx] + vars[idx]);
+            xhat[idx] = xhatMinus[idx] + k[idx]*(fluxValues[idx]-xhatMinus[idx]);
+            p[idx] = (1.0-k[idx])*pMinus[idx];
+        }
+    }
+    
+    return std::make_tuple(xhat, p);
+}
+    
+/* Apply kalman filtering to a single index 
+   and return the new likelihood */
+std::tuple<std::vector<int>, double> kalmanFilterIndex(std::vector<double> psiCurve,
+                                                       std::vector<double> phiCurve)
+    
+{
+    int numValues = psiCurve.size();
+    double maskVal = 1.0 / 9999999.0;
+    std::vector<double> fluxValues(numValues, 0.0);
+    std::vector<double> vars(numValues, maskVal);
+    std::vector<int> fluxIdx;
+    
+    /* Return a set of curve indicies equal to
+       { -1 } to indicate that this index should
+       be filtered out. */ 
+    std::vector<int> failIndex = { -1 };
+    auto filterIndex = std::make_tuple(failIndex, 0.0);
+    
+    // Mask phi values, calculate the flux,
+    // the inverse phi values, and get a list
+    // of positive flux values.
+    for(int i = 0; i < numValues; i++) {
+        double masked_phi = phiCurve[i];
+        if(masked_phi == 0.0 ) {
+            masked_phi = 1e9;
+        }
+        if(masked_phi < -999.0) {
+            masked_phi = 9999999.0;
```

Comment:
Are the values of `masked_phi` on line 154 and 157 magic numbers, or are they just suppose to be very large?

Reply:
I think that they might just be large numbers? I kept the values the same as the ones in the python implementation but it might not actually make a difference 🤷‍♀️